Purchases: Refactor cancel purchase page to use new layout

ciprianimike opened the pull request:
#### Proposed Changes

Restyle Cancel Purchase page, using a new layout with a right sidebar and a list of features for WPCom plan. 
P2: pebzTe-aB-p2

#### Design
![image](https://user-images.githubusercontent.com/5706607/206586127-b60b81cf-1b76-47e6-9d98-a3d5e8855ab5.png)

#### Before
<img width="1632" alt="Screenshot 2022-12-09 at 00 15 42" src="https://user-images.githubusercontent.com/5706607/206587017-7ac55c5a-f719-40d5-8331-97b2b154fa0a.png">
<img width="373" alt="Screenshot 2022-12-09 at 00 15 56" src="https://user-images.githubusercontent.com/5706607/206587029-59c3533d-2f6e-4cac-bb71-5cc6d11666f5.png">

#### After
<img width="1633" alt="Screenshot 2022-12-09 at 00 16 44" src="https://user-images.githubusercontent.com/5706607/206587063-65d87f9e-1196-4977-97ba-fee8bc6f8ef1.png">
<img width="374" alt="Screenshot 2022-12-09 at 00 16 31" src="https://user-images.githubusercontent.com/5706607/206587071-3c9d1e4a-1180-45e6-9e20-8561a366d2f5.png">

#### Specific cases

The following are specific cases, for which the page layout changes, as they make use of the usual CancelPurchase component.

## Domains
<img width="1501" alt="case Domain" src="https://user-images.githubusercontent.com/5706607/206710592-11366fd0-9822-4cf3-8c96-582f615d0ba7.png">

## Jetpack products
<img width="1501" alt="case Jepack" src="https://user-images.githubusercontent.com/5706607/206710539-9743369e-1d4d-41f4-b426-c35e0e151ac3.png">

## One-time purchases
<img width="1501" alt="case One Time Purchase" src="https://user-images.githubusercontent.com/5706607/206710567-ed9f2af8-1b5a-4a3c-a19b-8885f694e805.png">

#### Testing Instructions

1. Visit <Calypso>/me/purchases
2. Select a purchase
3. Cancel the purchase
5. Verify the page layout is changed.
6. Verify the Loading Placeholder is updated with the new design
7. Verify the Feature Set for WPCom Plans are properly shown
8. Verify that the copy for Subscriptions accordingly with the design
9. Verify the cancel functionality works properly
10. Verify the responsive behaviour
11. The following are specific cases (see Specific cases section):
    - Domains
    - Jetpack products
    - One-Time purchases

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

ciprianimike committed:
add cancelingSubscription to titles

ciprianimike committed:
style, page header and back link

ciprianimike committed:
show the new page for subscriptions only

ciprianimike committed:
duplicate refund-confirmation component and styles

southp commented:
@ciprianimike thanks for opening up the PR and the technical discussion early.

Comparing the new design:
![image](https://user-images.githubusercontent.com/1842898/206082514-084d92ab-be85-4d05-ac2d-eaf75cd2388f.png)

and what we have now, we can see that we already have all the data we need, except for the feature list . From the technical point of view, this page only consists of common components and its own components that aren't used in anywhere else. Thus, this PR should mainly about 2 things only:

1. Restructure the existing component into the new design. It's expectable that there might be things we need to breakdown. 
2. Embed [the cancellation feature list](https://github.com/Automattic/wp-calypso/pull/69682) and restyle accordingly. We'd want to migrate the code since we won't do the precancellation modal, but that can be a separate PR. From a glance it can be a component embedded into `<CancelPurchaseRefundInformation/>, but I'll defer that to your best judgement. 

The current approach is essentially a new page when it is a plan subscription; the new <CancelPurchaseRefundInformationWithFeatures /> is also mostly duplicated from <CancelPurchaseInformation>. Since it's a full-rollout, rather than guarding by a feature flag or an A/B test, let's aim for restructuring the current code into its new form :)  

Once the above foundational work is settled, we can the fine-tune the whole page. e.g. 
* For the background color you mentioned to me, we can likely refer to how some pages done it. e.g. /domains/manage. 
* Note that we'd need to update the loading placeholder as well, which is based on the old design and is quite off combining with the new design:
<img width="974" alt="image" src="https://user-images.githubusercontent.com/1842898/206083675-c0ba38ec-7287-42d7-9b63-fe616e4c5b01.png">

ciprianimike commented:
> @ciprianimike thanks for opening up the PR and the technical discussion early.
> 
> Comparing the new design: ![image](https://user-images.githubusercontent.com/1842898/206082514-084d92ab-be85-4d05-ac2d-eaf75cd2388f.png)
> 
> and what we have now, we can see that we already have all the data we need, except for the feature list . From the technical point of view, this page only consists of common components and its own components that aren't used in anywhere else. Thus, this PR should mainly about 2 things only:
> 
> 1. Restructure the existing component into the new design. It's expectable that there might be things we need to breakdown.
> 2. Embed [the cancellation feature list](https://github.com/Automattic/wp-calypso/pull/69682) and restyle accordingly. We'd want to migrate the code since we won't do the precancellation modal, but that can be a separate PR. From a glance it can be a component embedded into `, but I'll defer that to your best judgement.
> 
> The current approach is essentially a new page when it is a plan subscription; the new is also mostly duplicated from . Since it's a full-rollout, rather than guarding by a feature flag or an A/B test, let's aim for restructuring the current code into its new form :)
> 
> Once the above foundational work is settled, we can the fine-tune the whole page. e.g.
> 
> * For the background color you mentioned to me, we can likely refer to how some pages done it. e.g. /domains/manage.
> * Note that we'd need to update the loading placeholder as well, which is based on the old design and is quite off combining with the new design:
> 
> <img alt="image" width="974" src="https://user-images.githubusercontent.com/1842898/206083675-c0ba38ec-7287-42d7-9b63-fe616e4c5b01.png">

Initially I thought the same as your comment, then I duplicated the component because with [this PR](https://github.com/Automattic/wp-calypso/pull/70493/files#diff-811c6e3fd087924c069c3a6a662ecf053f21649dabd23857e6d6c431e0d7a17f), Jetpack subscriptions were added to the page. If I restyle the page without discriminating, the Jetpack flow will be impacted too and we have no design for that.

The component CancelPurchaseRefundInformation has different elements and copy compared to the Design received.

The loading placeholder should be conditional too, showing the old page structure if `isJetpackProduct`.

ciprianimike committed:
add feature list

ciprianimike committed:
add feature list

ciprianimike committed:
feature list CSS tuning

ciprianimike committed:
CSS footer buttons

ciprianimike committed:
fix page background

ciprianimike commented:
@southp I fixed the background page and added feature lists, on the base of the Jetpack case I'll need to add a check [in the controller](https://github.com/Automattic/wp-calypso/blob/99e581f02c3f16373e30154307a8eb660b051893/client/me/purchases/controller.jsx#L88).

ciprianimike committed:
refactor refund-information component

ciprianimike committed:
site header css breakpoint

ciprianimike committed:
refund-information component responsive tuning

ciprianimike committed:
refund-information form style

ciprianimike committed:
case refund+domain

ciprianimike committed:
Cancel cases copy

ciprianimike committed:
cancel purchase loading placeholder

ciprianimike committed:
refactor cancel purchase loading placeholder

ciprianimike committed:
cancel purchase loading purchase responsive CSS

southp commented:
@ciprianimike Could you help me confirm if you have anything uncommitted? This is what I'm seeing:
<img width="1255" alt="image" src="https://user-images.githubusercontent.com/1842898/206635142-95092da3-6dd8-4332-a59c-d9915a30ec27.png">

I've also checked that I'm at the latest tip and have run `yarn run distclean && yarn install && yarn start` for a fresh start.
<img width="1194" alt="image" src="https://user-images.githubusercontent.com/1842898/206635228-329601ee-7d56-4839-9eac-40b655ceb58f.png">

ciprianimike committed:
fix desktop breakpoint

ciprianimike commented:
> @ciprianimike Could you help me confirm if you have anything uncommitted? This is what I'm seeing: <img alt="image" width="1255" src="https://user-images.githubusercontent.com/1842898/206635142-95092da3-6dd8-4332-a59c-d9915a30ec27.png">
> 
> I've also checked that I'm at the latest tip and have run `yarn run distclean && yarn install && yarn start` for a fresh start. <img alt="image" width="1194" src="https://user-images.githubusercontent.com/1842898/206635228-329601ee-7d56-4839-9eac-40b655ceb58f.png">

I reproduced this issue, I'm fixing it

ciprianimike committed:
fix /purchases/subscriptions/ cancel page

ciprianimike commented:
> @ciprianimike Could you help me confirm if you have anything uncommitted? This is what I'm seeing: <img alt="image" width="1255" src="https://user-images.githubusercontent.com/1842898/206635142-95092da3-6dd8-4332-a59c-d9915a30ec27.png">
> 
> I've also checked that I'm at the latest tip and have run `yarn run distclean && yarn install && yarn start` for a fresh start. <img alt="image" width="1194" src="https://user-images.githubusercontent.com/1842898/206635228-329601ee-7d56-4839-9eac-40b655ceb58f.png">

I pushed a CSS fix for the issue you were facing, you should now see the page like this:
<img width="1401" alt="Screenshot 2022-12-09 at 09 32 57" src="https://user-images.githubusercontent.com/5706607/206659494-f7f94508-3340-4d68-9ad9-806551e213eb.png">

dzver commented on client/me/purchases/cancel-purchase/loading-placeholder.jsx:
```diff
@@ -1,33 +1,89 @@
+
+	function siteHeader() {
+		return (
+			<div class="site">
```
I see some errors in the console
```
Warning: Invalid DOM property `class`. Did you mean `className`?
```

ciprianimike replied:
this is already fixed https://github.com/Automattic/wp-calypso/pull/70856/files#diff-60deb3bcb1a5e256394cc7630489a2e7ef67c7e48337190bb38884c06c9cd9cfR35 I suppose you tested before my last commit, but it fixed this and some CSS fine tuning. The other comments are still valid

dzver commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -149,29 +203,47 @@ const CancelPurchaseRefundInformation = ( {
+										</span>
+										<span className="cancel-budnled-domain__option-description">
+											{ i18n.translate(
+												"You'll receive a partial refund of %(planCost)s. The domain will be cancelled, you risk losaing it forever," +
```
I see two partial refund checkboxes, one is supposed to be a full refund, I think.

Do we need to change the copy for it? The old one is already translated.

ciprianimike replied:
I am working on the missing full refund case. The copy comes from the design (first screenshot) and this is the first case: `Plan in refund window with domain`

ciprianimike replied:
The partial refund issue has been fixed.
About the copy I'm implementing `hasTranslation( '<string>' ) || locale === 'en'` to all the new strings

dzver commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -149,29 +203,47 @@ const CancelPurchaseRefundInformation = ( {
-						</FormLabel>
+						</FormLabel>,
+						i18n.translate(
+							"Since you are cancelling your plan within %(refundPeriodInDays)d days of purchasing you'll receive a {{spanRefund}}refund of %(refundAmount)s{{/spanRefund}} and it will be removed from your site immediately.",
```
When I select the full refund other option, I still see the partial refund here (the green text at the bottom).

<img width="761" alt="Screenshot 2022-12-09 at 10 54 15" src="https://user-images.githubusercontent.com/82778/206663433-488b2c43-3bbd-4245-a492-0f77d95b4fbc.png">

ciprianimike replied:
this issue has been fixed, video attached
![Kapture 2022-12-09 at 15 10 52](https://user-images.githubusercontent.com/5706607/206721178-0e4e8604-134e-46e7-84a7-dc032b19de5c.gif)

dzver commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -215,30 +287,34 @@ const CancelPurchaseRefundInformation = ( {
-							'This plan includes the custom domain, %(domain)s, normally a %(domainCost)s purchase. ' +
-								'The domain will not be removed along with the plan, to avoid any interruptions for your visitors. ',
+							'This plan subscription includes the custom domain, %(domain)s' +
+								'{{strong}}The domain will not be removed{{/strong}} along with the plan, to avoid any interruptions for your visitors. ',
```
Does this change come from the design?

What's the plan for translations?

ciprianimike replied:
The copy comes from the design pebzTe-aB-p2#comment-876
I'm implementing `hasTranslation( '<string>' ) || locale === 'en'` to all the new strings

dzver reviewed (commented):
I did a round of reviews and left some feedback. A general observation - lots of copy seems to change. In case it comes from @JanaMW27 and has to be that way, perhaps either add a String Freeze tag or come up with another plan for the translations after talking to @southp 
 (for example, the new strings can be used for the English locale).

In case it doesn't come from @JanaMW27, perhaps in some cases, we could use the old copy.

I tried canceling plans, plan+domain, and a one-time product and didn't find anything major, just the things mentioned in the notes. Perhaps all plans need to be tested to confirm the features.

Ah, and this screen looks like it's missing a white background - is it on purpose:
<img width="1110" alt="Screenshot 2022-12-09 at 11 17 46" src="https://user-images.githubusercontent.com/82778/206667716-1f8d2319-e08b-4aa2-b863-af1a9e68c502.png">

ciprianimike committed:
remove comment

ciprianimike committed:
loading placeholder mobile css fix

ciprianimike commented:
> Ah, and this screen looks like it's missing a white background - is it on purpose: <img alt="Screenshot 2022-12-09 at 11 17 46" width="1110" src="https://user-images.githubusercontent.com/82778/206667716-1f8d2319-e08b-4aa2-b863-af1a9e68c502.png">

The screenshot attached shows a page with the white background, anyhow I fixed a case for that problem with the comment, https://github.com/Automattic/wp-calypso/pull/70856#issuecomment-1344006532

as for the className react warning, I support your tests started before I pushed last tuning, it should be already fixed.

ciprianimike committed:
fix typo

ciprianimike committed:
re-enable eslint via comment

ciprianimike committed:
full refund fix

southp commented:
Normally I'd submit all the feedback at once, but I'll just comment whatever I found immediately so we can work in parallel for a better throughput :) 

To begin with, the date seems missing:
<img width="617" alt="image" src="https://user-images.githubusercontent.com/1842898/206729840-6eff4ece-c39b-4e86-a10f-3af998c9399f.png">

southp commented:
On cancelling the Business plan, "And more" now looks quite off on the grid layout:
<img width="958" alt="image" src="https://user-images.githubusercontent.com/1842898/206730908-83ec6621-de48-4181-8a06-236ffdf4ec7f.png">
Could we make it more natural in this new layout? @JanaMW27 how would you recommend?

ciprianimike commented:
> On cancelling the Business plan, "And more" now looks quite off on the grid layout: <img alt="image" width="958" src="https://user-images.githubusercontent.com/1842898/206730908-83ec6621-de48-4181-8a06-236ffdf4ec7f.png"> Could we make it more natural in this new layout? @JanaMW27 how would you recommend?

I removed the X icon since it was requested in the previous use of the cancellation features, I can bring it back, the design shows it.

southp commented:
> @southp I removed the X icon since it was requested in the previous use of the cancellation features, I can bring it back, the design shows it.

Yeah, I personally would still wonder [whether we should show the X icon on a non-feature thing](https://github.com/Automattic/wp-calypso/pull/69682#discussion_r1013179728). However, it seems make better sense on this grid layout. Let's bring it back.

ciprianimike committed:
fix className missing

southp commented:
On cancelling the Premium plan with a bundled domain that is outside of the refund window, there is no space between the domain name and the following copy:
<img width="945" alt="image" src="https://user-images.githubusercontent.com/1842898/206733791-ca5763fa-651d-49b5-bb4a-3e51d0070f80.png">

ciprianimike committed:
add support for hasTranslation and locale checks

ciprianimike committed:
remove commented code

southp commented:
There should be a space following the comma:
<img width="655" alt="image" src="https://user-images.githubusercontent.com/1842898/206735750-b3bdf3c9-78df-4624-b575-943a16073447.png">

ciprianimike committed:
translation fallbacks in CancelPurchase component

ciprianimike committed:
remove unused title, added previously in this PR

ciprianimike committed:
remove incorrect header

ciprianimike committed:
remove wrong sentence

ciprianimike commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -43,6 +46,60 @@ const CancelPurchaseRefundInformation = ( {
+			<div>
+				<p className="cancel-purchase__refund-information-subtitle">
+					{ i18n.translate(
+						'If you cancel your plan subscription your site may appear broken and things may not work properly.'
```
@dzver  @southp  we don't have a valid fallback for this string and it's the copy header for the feature set
<img width="728" alt="Screenshot 2022-12-09 at 16 58 41" src="https://user-images.githubusercontent.com/5706607/206742222-0916ef81-0ea9-4d36-a337-73be9a682bae.png">

southp replied:
Let's add String Freeze for now, since we have to wait for other things anyway.

ciprianimike replied:
string freeze was added

southp commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -149,29 +203,47 @@ const CancelPurchaseRefundInformation = ( {
+												}
+											) }
+										</span>
+										<span className="cancel-budnled-domain__option-description">
```
Typo: "budnled" should be "bundled"

ciprianimike replied:
typo fixed

southp commented on client/me/purchases/cancel-purchase/loading-placeholder.jsx:
```diff
@@ -1,33 +1,94 @@
 		path = getManagePurchaseUrlFor( siteSlug, purchaseId );
 	}
 
+	function goBack() {
```
I believe what @sirbrillig [commented previously](https://github.com/Automattic/wp-calypso/pull/69276#discussion_r1004638963) also applies here: the actual definition of`goBack()`, `supportLink()` and `siteHeader()` will change everytime `CancelPurchaseLoadingPlaceholder` is called. In general, if a lambda function won't change based on the runtime context of the function it resides in, most of the time it'd better to be extracted. The end result also tends to be more readable.

There are some other similar cases throughout this PR that worth considering this. I'll defer that to you :)

ciprianimike replied:
Fixed, I'll avoid reusing functions like this

ciprianimike replied:
Fixed, I'll avoid to use functions in this way

southp commented on client/me/purchases/cancel-purchase/index.jsx:
```diff
@@ -209,49 +229,113 @@ class CancelPurchase extends Component {
+										{ this.renderFooterText( this.props ) }
+									</div>
+
+									<CancelPurchaseButton
```
Now that this button is in the `! isSubscription()` condition, I'm wondering if it can be simpler. Most of the props passed to it clearly doesn't make sense under this condition, e.g. `cancelBundledDomain`

ciprianimike replied:
cancelBundledDomain arrives from the props and is mandatory for the <CancelPurchaseButton> mandatory, to simplify I removed the `disabled` attribute since it doesn't make sense here, but let me know if you want me to refine the logic in the <CancelPurchaseButton> component

southp commented on client/me/purchases/cancel-purchase/get-plan-cancellation-features.jsx:
```diff
@@ -0,0 +1,57 @@
+import { getPlan, isMonthly } from '@automattic/calypso-products';
```
This file is simply a jsx version of client/me/purchases/pre-cancellation-dialog/get-plan-cancellation-features.tsx. We should either remove the tsx file here, or add a TODO note making it clear that we will remove the tsx version as a follow-up. Otherwise we risk making other developers confused about why there are two seemingly identical files.

ciprianimike replied:
My idea was to remove the TSX file in PR to clean up the work done for pre-cancellation modal. Where should I add the TODO?

southp replied:
The beginning of this file will be good enough :)

ciprianimike replied:
done

sirbrillig commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -43,6 +46,67 @@ const CancelPurchaseRefundInformation = ( {
 		} );
 	};
 
+	const featuresList = ( productSlug, hasDomain ) => {
```
Similar to one of the comments above, this function is effectively a component: it's a function which accepts arguments (props) and returns HTML.

Could this be moved into its own component? React will do a better job of optimising it, its dependencies will be more clear, and it will be more readable.

ciprianimike replied:
`featuresList` has been move outside the main component

sirbrillig commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -43,6 +46,67 @@ const CancelPurchaseRefundInformation = ( {
+		);
+	};
+
+	const cancelSubscriptionDescription = () => {
```
Same comment here. `cancelSubscriptionDescription` is effectively a component so let's just move it out of this function.

ciprianimike replied:
`cancelSubscriptionDescription` has been move outside the main component

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7664608

Thank you @ciprianimike for including a screenshot in the description! This is really helpful for our translators.

sirbrillig commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -43,6 +46,67 @@ const CancelPurchaseRefundInformation = ( {
+						>
+							<span className="cancel-purchase__refund-information--item-more-span">
+								{ i18n.translate( 'And more' ) }
+							</span>
```
It's probably too much for this PR since they seem to already be scattered around this file but it's better to use `translate` from the `useTranslate` hook rather than calling it directly. I can't remember all the reasons for this but I think that one of them is that if the locale changes, updating the provider will not update these translations.

ciprianimike replied:
I switched to `useTranslate` in all `cancel-purchase/` files

sirbrillig commented on client/me/purchases/cancel-purchase/index.jsx:
```diff
@@ -184,19 +196,27 @@ class CancelPurchase extends Component {
 		}
 
 		const { purchase, isJetpackPurchase } = this.props;
+		const locale = getCurrentLocaleSlug();
```
`getCurrentLocaleSlug()` is a Redux selector and requires passing in the state, so this will always return `null`. You need to call this in the `mapStateToProps` function at the bottom of this file.

ciprianimike replied:
`getCurrentLocaleSlug()` has been moved to the mapStateToProps, not sure why it returned the language to me

sirbrillig commented on client/me/purchases/cancel-purchase/index.jsx:
```diff
@@ -209,49 +229,113 @@ class CancelPurchase extends Component {
+						<CancelPurchaseRefundInformation
+							purchase={ purchase }
+							isJetpackPurchase={ isJetpackPurchase }
+							includedDomainPurchase={ this.props.includedDomainPurchase }
```
`includedDomainPurchase` is not a prop that can be passed to `CancelPurchaseRefundInformation`; it's automatically set by the `mapStateToProps` function in that component based on the `purchase` prop. So this line should be removed or (if providing it is necessary), `CancelPurchaseRefundInformation` should be modified to allow setting it as a prop.

ciprianimike replied:
`includedDomainPurchase`  has been removed as passed parameter for `<CancelPurchaseRefundInformation>`

sirbrillig commented on client/me/purchases/cancel-purchase/index.jsx:
```diff
@@ -209,49 +229,113 @@ class CancelPurchase extends Component {
+
+								<CompactCard className="cancel-purchase__footer">
+									<div className="cancel-purchase__refund-amount">
+										{ this.renderFooterText( this.props ) }
```
`renderFooterText` does not accept any arguments, although arguably you could make it into a functional component of its own.

ciprianimike replied:
renderFooterText has been moved outside the main component

sirbrillig commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -284,45 +360,61 @@ const CancelPurchaseRefundInformation = ( {
-					mappingCost: includedDomainPurchase.priceText,
-				},
-			}
+		text.push(
```
I think that `text` is undefined here, so this call to `push()` will result in a fatal error.

ciprianimike replied:
issue has been fixed

sirbrillig replied:
Nice!

sirbrillig reviewed (commented):
I noted several issues inline, some of them significant. It's worth noting that one way I found them was by adding `// @ts-check` to the top of the JSX files temporarily and allowing the type checker to look for problems. This is a helpful technique if you are not using typescript.

DavidRothstein commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -43,6 +46,67 @@ const CancelPurchaseRefundInformation = ( {
+		return (
+			<Fragment>
+				<div className="cancel-purchase__refund-information-description-header">
+					{ i18n.translate( 'When your %(productName)s plan expires you will loose access to:', {
```
Should be "lose", not "loose".  Also probably a comma after "expires".

Also, it talks about "when your plan expires" but isn't this sentence also being displayed in the cancel-with-refund flow, when the plan will be removed immediately after canceling?  (I didn't test, but that's my sense from the screenshots.)

ciprianimike replied:
I fixed the copy, do you want me to remove this string instead?

DavidRothstein replied:
Thanks!  I wouldn't necessarily remove it personally, just think that the wording doesn't really work for the case where the plan is being removed immediately (it does work fine for the non-refundable case where the cancellation just turns auto-renew off).  I'm not sure, but maybe something like "When your plan is removed" or "When your plan is cancelled" would work for the first case?

DavidRothstein commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -43,6 +46,67 @@ const CancelPurchaseRefundInformation = ( {
+			<div>
+				<p className="cancel-purchase__refund-information-subtitle">
+					{ i18n.translate(
+						'If you cancel your plan subscription your site may appear broken and things may not work properly.'
```
"things" does not sound very professional here, and the whole sentence is kind of wishy-washy.... In general I feel like we reserve this kind of language for cases where we know something is likely to undergo major changes (site will lose its current premium theme, Atomic site will be reverted, etc).

Could maybe say something simpler like "your site will lose functionality and its appearance may change", but overall -- does this sentence even add anything useful, given the more specific list of things that will be lost which is displayed right underneath it?

ciprianimike replied:
I updated the copy, LMK if you want me to remove this text

DavidRothstein replied:
I'm not sure what others think, but personally that's what I would do.

DavidRothstein commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -95,50 +151,48 @@ const CancelPurchaseRefundInformation = ( {
-							'Your plan included the custom domain %(domain)s. You can cancel your domain as well as the plan, but keep ' +
-								'in mind that when you cancel a domain you risk losing it forever, and visitors to your site may ' +
-								'experience difficulties accessing it.',
+							'Your plan includes the custom domain “%(domain)s”. What would you like to do with the domain?',
```
That is a nice improvement to the text :+1:

DavidRothstein commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -149,29 +203,47 @@ const CancelPurchaseRefundInformation = ( {
+										<span className="cancel-budnled-domain__option-description">
+											{ i18n.translate(
+												"You'll receive a full refund of %(planCost)s. The domain will be cancelled, you risk losing it forever," +
+													'and visitors to your site may experience difficulties accessing it.',
```
Probably "difficulty" rather than "difficulties"?

ciprianimike replied:
copy updated

DavidRothstein commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -149,29 +203,47 @@ const CancelPurchaseRefundInformation = ( {
-						</FormLabel>
+						</FormLabel>,
+						i18n.translate(
+							"Since you are cancelling your plan within %(refundPeriodInDays)d days of purchasing you'll receive a {{spanRefund}}refund of %(refundAmount)s{{/spanRefund}} and it will be removed from your site immediately.",
```
I don't know if we're ever likely to have a 1-day refund period, but this translation should probably handle the singular-vs.-plural case anyway (my understanding is also that there are some languages where the singular form is used even when the number isn't 1 -- see https://codex.wordpress.org/I18n_for_WordPress_Developers#Plurals).

I would also put a comma after "purchasing".

ciprianimike replied:
I added singular vs plural and the comma

DavidRothstein commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -215,30 +287,34 @@ const CancelPurchaseRefundInformation = ( {
-							'This plan includes the custom domain, %(domain)s, normally a %(domainCost)s purchase. ' +
-								'The domain will not be removed along with the plan, to avoid any interruptions for your visitors. ',
+							'This plan subscription includes the custom domain, %(domain)s' +
+								'{{strong}}The domain will not be removed{{/strong}} along with the plan, to avoid any interruptions for your visitors. ',
```
Is this missing a period between the two sentences?

ciprianimike replied:
Copy updated

DavidRothstein commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -284,45 +360,61 @@ const CancelPurchaseRefundInformation = ( {
+		text.push(
+			i18n.translate(
+				'This plan subscription includes the custom domain, %(domain)s' +
+					'{{strong}}The domain will not be removed{{/strong}} along with the plan, to avoid any interruptions for your visitors. ',
```
Is this missing a period between the two sentences?

ciprianimike replied:
Copy updated

DavidRothstein commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -284,45 +360,61 @@ const CancelPurchaseRefundInformation = ( {
+				}
+			),
+			i18n.translate(
+				'If you cancel your plan subscription your plan will be removed on %(expireDate)s.',
```
(minor) Comma after "subscription", probably.

ciprianimike replied:
comma added

DavidRothstein commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -284,45 +360,61 @@ const CancelPurchaseRefundInformation = ( {
-			'This plan includes the custom domain, %(domain)s. ' +
-				'The domain will not be removed along with the plan, to avoid any interruptions for your visitors. ',
+			'This plan subscription includes the custom domain, %(domain)s' +
+				'{{strong}}The domain will not be removed{{/strong}} along with the plan, to avoid any interruptions for your visitors. ',
```
Is this missing a period between the two sentences?

DavidRothstein commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -284,45 +360,61 @@ const CancelPurchaseRefundInformation = ( {
 		text = i18n.translate(
-			"When you cancel your subscription, you'll be able to use %(productName)s until your subscription expires. " +
-				'Once it expires, it will be automatically removed from your site.',
+			'If you cancel your plan subscription your plan will be removed on %(expireDate)s.',
```
(minor) Comma after "subscription", probably.

ciprianimike replied:
comma added

DavidRothstein reviewed (commented):
This is not a real review of the actual code, but I noticed a number of typos and other issues with the wording.  I commented on the ones I saw during a quick read-through, but this probably needs some proofreading and overall editorial review given the large number of changes it's making to the user-facing text?

ciprianimike committed:
copy fixes

ciprianimike committed:
fix internal function JS performance issue

ciprianimike committed:
remove disable state for CancelPurchaseButton subscriptions with no plan

ciprianimike committed:
LoadingPlaceholder use useTranslate on i18n.translate

ciprianimike commented:
> Normally I'd submit all the feedback at once, but I'll just comment whatever I found immediately so we can work in parallel for a better throughput :)
> 
> To begin with, the date seems missing: <img alt="image" width="617" src="https://user-images.githubusercontent.com/1842898/206729840-6eff4ece-c39b-4e86-a10f-3af998c9399f.png">

this issue has been fixed

ciprianimike committed:
bring back the X icon in the 'add more' item

ciprianimike commented:
> > @southp I removed the X icon since it was requested in the previous use of the cancellation features, I can bring it back, the design shows it.
> 
> Yeah, I personally would still wonder [whether we should show the X icon on a non-feature thing](https://github.com/Automattic/wp-calypso/pull/69682#discussion_r1013179728). However, it seems make better sense on this grid layout. Let's bring it back.

X icon has been brought back for "And more" item

ciprianimike commented:
> There should be a space following the comma:

issue has been fixed

ciprianimike committed:
refund-information use useTranslate vs i18n.translate

ciprianimike committed:
use getCurrentLocale in props map

ciprianimike committed:
move components outsite main component

ciprianimike committed:
fix expiryDate missing

emilyaudela commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -1,21 +1,83 @@
+	return (
+		<Fragment>
+			<div className="cancel-purchase__refund-information-description-header">
+				{ translate( 'When your %(productName)s plan expires you will loose access to:', {
```
There is a typo in this string. `loose` -> `lose`

ciprianimike replied:
this typo has been fixed

ciprianimike committed:
fix typo

ciprianimike committed:
add todo to remove pre-cancellation modal get-plan-cancellation-features

ciprianimike committed:
make renderFooterText an external component

ciprianimike committed:
remove props.includedDomainPurchase fro <CancelPurchaseButton>

ciprianimike committed:
bring 'includedDomainPurchase' back to <CancelPurchaseButton>

ciprianimike committed:
fix text undefined variable

ciprianimike committed:
remove ts-check

ciprianimike committed:
copy: add comma

ciprianimike committed:
Copy: 'difficulty' rather than 'difficulties'

ciprianimike committed:
update copy

ciprianimike committed:
add singular vs plural to trasnlate() for refundPeriodInDays

ciprianimike committed:
add period and <br />

ciprianimike committed:
copy: add comma

ciprianimike committed:
fix period

ciprianimike committed:
fix period

ciprianimike committed:
add ! isRefundable Jetpack

sirbrillig commented on client/me/purchases/cancel-purchase/loading-placeholder.jsx:
```diff
@@ -1,33 +1,93 @@
 
+import '../components/loading-placeholder/style.scss';
+
+const supportLink = ( translate ) => {
```
Minor: there's nothing wrong with creating a plain function like this and `siteHeader` to return HTML, but they are effectively React functional components, so the common pattern is to do a few things differently:

1. Capitalize the function name. In this case, `supportLink` becomes `SupportLink`.
2. Change its arguments to an object. In this case `(translate)` becomes `({translate})`. Alternatively since you can use hooks inside the function so you could remove this argument and on the next line write `const translate = useTranslate()`.
3. Change the call site to JSX. In this case, `{ supportLink( translate ) }` becomes `<SupportLink />`.

ciprianimike replied:
Thanks for the explanation, I addressed it. I'll use the approach indicated

sirbrillig commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -1,33 +1,100 @@
 import { getIncludedDomainPurchase } from 'calypso/state/purchases/selectors';
+import getPlanCancellationFeatures from './get-plan-cancellation-features';
+
+const featuresList = ( translate, purchase, productSlug, hasDomain ) => {
```
This looks like a great change for readability! As above, however, this really feels like it should be a component. If you change it to `FeaturesList` and its parameters to `( { translate, purchase, productSlug, hasDomain } )` and the call site to JSX, that's all you would need!

ciprianimike replied:
FeatureList is now a React component

sirbrillig commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -1,33 +1,100 @@
+	);
+};
+
+const cancelSubscriptionDescription = ( site, translate, purchase, includedDomainPurchase ) => {
```
Same here, this could be `CancelSubscriptionDescription`.

ciprianimike replied:
CancelSubscriptionDescription is now a React component

sirbrillig commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -45,29 +112,23 @@ const CancelPurchaseRefundInformation = ( {
+
 			if ( includedDomainPurchase && isDomainMapping( includedDomainPurchase ) ) {
 				text.push(
-					i18n.translate(
```
Thank you for changing these! That will definitely help the future of this file.

sirbrillig commented on client/me/purchases/cancel-purchase/index.jsx:
```diff
@@ -129,42 +165,35 @@ class CancelPurchase extends Component {
+	keepButtonCTA = ( locale ) => {
+		let cta;
+
+		if ( isJetpackProduct ) {
```
Since `isJetpackProduct` is a function imported from `@automattic/calypso-products`, this will always be true. Did you mean to check `this.props.isJetpackPurchase`?

ciprianimike replied:
Yes I meant `this.props.isJetpackPurchase`, I fixed this issue

sirbrillig commented on client/me/purchases/cancel-purchase/index.jsx:
```diff
@@ -209,49 +245,113 @@ class CancelPurchase extends Component {
+						<CancelPurchaseRefundInformation
+							purchase={ purchase }
+							isJetpackPurchase={ isJetpackPurchase }
+							confirmBundledDomain={ this.state.confirmCancelBundledDomain }
```
There's no `confirmBundledDomain` prop on `CancelPurchaseRefundInformation`. Did you mean `confirmCancelBundledDomain`?

ciprianimike replied:
`confirmBundledDomain`  has been replaced with replaced with `confirmCancelBundledDomain`

sirbrillig replied:
It hasn't, though. The code here still reads,

```
confirmBundledDomain={ this.state.confirmCancelBundledDomain }
```

But that prop, `confirmBundledDomain`, does not exist.

ciprianimike replied:
I double checked it, it's fixed now, sorry

sirbrillig commented on client/me/purchases/cancel-purchase/index.jsx:
```diff
@@ -209,49 +245,113 @@ class CancelPurchase extends Component {
+											this.props.translate,
+											this.props.moment,
+											this.includedDomainPurchase,
+											purchase
```
I don't see a `includedDomainPurchase` property in this component. Did you mean `this.props.includedDomainPurchase`?

ciprianimike replied:
`includedDomainPurchase` has been replaced with `this.props.includedDomainPurchase`

sirbrillig reviewed (commented):
Thanks for the fixes! I noticed a couple more issues which I've noted inline.

ciprianimike committed:
supportLink() as React component

ciprianimike committed:
featureList() as React component

ciprianimike committed:
refund-feature fix form labels copy

ciprianimike committed:
cancelSubscriptionDescription() as React component

ciprianimike committed:
fix isJetpackProduct vs this.props.isJetpackPurchase

ciprianimike committed:
fix props on CancelPurchaseRefundInformation component

ciprianimike committed:
use includedDomainPurchase as prop

ciprianimike committed:
fix FooterText component

sirbrillig commented on client/me/purchases/cancel-purchase/loading-placeholder.jsx:
```diff
@@ -1,33 +1,94 @@
+	} );
+};
+
+const siteHeader = () => {
```
Minor: this could also be a component: `<SiteHeader />`

ciprianimike replied:
SiteHeader is now a react component

ciprianimike committed:
siteHeader as React component

ciprianimike committed:
fix confirmCancelBundledDomain in CancelPurchaseRefundInformation component

sirbrillig reviewed (approved):
I didn't test this (I leave that to your team) but from a code perspective this looks good to me!

DavidRothstein commented on client/me/purchases/cancel-purchase/refund-information.jsx:
```diff
@@ -149,34 +215,62 @@ const CancelPurchaseRefundInformation = ( {
+						) +
+							' ' +
+							translate(
+								"you'll receive a {{spanRefund}}refund of %(refundAmount)s{{/spanRefund}} and it will be removed from your site immediately.",
```
I saw the following on my screen here:

>Since you are cancelling your plan within 14 days of purchasing, [object Object]

I also don't think that translatable strings should be concatenated mid-sentence like this.  Shouldn't the whole sentence be translated as a single string?

ciprianimike replied:
I replicated and fixed it.
<img width="1045" alt="Screenshot 2022-12-15 at 12 10 25" src="https://user-images.githubusercontent.com/5706607/207844704-cb8b021d-45a2-4aa9-9590-e24f87320ab3.png">

ciprianimike committed:
fix concatenated translated strings

ciprianimike committed:
use quotes with domain name in all cases

JanaMW27 commented:
One little comment for this case:
<img width="820" alt="Screenshot 2022-12-20 at 08 25 50" src="https://user-images.githubusercontent.com/71137829/208607955-c905399f-1ed7-4dde-ac1c-4c3676d1aa63.png">

We should include the sentence that says that the plan will remove the date date date.

It is not a blocker, but maybe we can do a little fix once this is released?

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7664608).

DavidRothstein commented:
I noticed a number of problems here, including a couple already discussed in reviews above but also a few new ones (two of them pretty bad).  For the moment, I've documented everything I found in one issue: https://github.com/Automattic/wp-calypso/issues/71577

sirbrillig commented:
For the record, this has been reimplemented and merged in https://github.com/Automattic/wp-calypso/pull/95065 (developed in https://github.com/Automattic/wp-calypso/pull/93211)
